server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
+            }
+        }
+
+        logger.info(() -> new ParameterizedMessage("adding block {} to indices {}", block.name,
+            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.joining(","))));
+        return ClusterState.builder(currentState).blocks(blocks).metadata(metadata).routingTable(routingTable.build()).build();
+    }
+
+    public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request,
+                              ActionListener<AddIndexBlockResponse> listener) {
+        final Index[] concreteIndices = request.indices();
+        if (concreteIndices == null || concreteIndices.length == 0) {
+            throw new IllegalArgumentException("Index name is required");
+        }
+        List<String> writeIndices = new ArrayList<>();
+        SortedMap<String, IndexAbstraction> lookup = clusterService.state().metadata().getIndicesLookup();
+        for (Index index : concreteIndices) {
+            IndexAbstraction ia = lookup.get(index.getName());
+            if (ia != null && ia.getParentDataStream() != null && ia.getParentDataStream().getWriteIndex().getIndex().equals(index)) {
+                writeIndices.add(index.getName());
+            }
+        }
+        if (writeIndices.size() > 0) {
+            throw new IllegalArgumentException("cannot add a block to the following data stream write indices [" +
+                Strings.collectionToCommaDelimitedString(writeIndices) + "]");
+        }
+
+        clusterService.submitStateUpdateTask("add-index-block " + Arrays.toString(concreteIndices),
+            new ClusterStateUpdateTask(Priority.URGENT) {
+
+                private final Map<Index, ClusterBlock> blockedIndices = new HashMap<>();
+
+                @Override
+                public ClusterState execute(final ClusterState currentState) {
+                    return addIndexBlock(concreteIndices, blockedIndices, currentState, request.getBlock());
+                }
+
+                @Override
+                public void clusterStateProcessed(final String source, final ClusterState oldState, final ClusterState newState) {
+                    if (oldState == newState) {
+                        assert blockedIndices.isEmpty() : "List of blocked indices is not empty but cluster state wasn't changed";
+                        listener.onResponse(new AddIndexBlockResponse(true, false, Collections.emptyList()));
+                    } else {
+                        assert blockedIndices.isEmpty() == false : "List of blocked indices is empty but cluster state was changed";
+                        threadPool.executor(ThreadPool.Names.MANAGEMENT)
+                            .execute(new WaitForBlocksApplied(blockedIndices, request,
+                                    ActionListener.wrap(verifyResults ->
+                                            clusterService.submitStateUpdateTask("finalize-index-block",
```

Comment:
Can you add the block type and `blockedIndices` list to this state update task as well?